torchrl/objectives/costs/dreamer.py

```diff
@@ -0,0 +1,175 @@
+            tensordict["posterior_means"],
+            tensordict["posterior_stds"],
+        )
+        reco_loss = (
+            0.5
+            * self.reco_loss(
+                tensordict.get("pixels"),
+                tensordict.get("reco_pixels"),
+            )
+            .mean(dim=[0, 1])
+            .sum()
+        )
+        reward_loss = 0.5 * self.reward_loss(
+            tensordict.get("reward"), tensordict.get("predicted_reward")
+        )
+        loss = (
+            self.lambda_kl * kl_loss
+            + self.lambda_reco * reco_loss
+            + self.lambda_reward * reward_loss
+        )
+        return (
+            TensorDict(
+                {
+                    "loss_world_model": loss,
+                    "loss_kl": kl_loss,
+                    "loss_reco": reco_loss,
+                    "loss_reward": reward_loss,
+                },
+                [],
+            ),
+            tensordict,
+        )
+
+    def kl_loss(self, prior_mean, prior_std, posterior_mean, posterior_std):
+        flat_prior = d.Normal(prior_mean, prior_std)
+        flat_posterior = d.Normal(posterior_mean, posterior_std)
+        kl = kl_divergence(flat_posterior, flat_prior)
+        kl = kl.clamp(min=self.free_nats).mean()
+        return kl
+
+
+class DreamerBehaviourLoss(LossModule):
+    def __init__(
+        self,
+        actor_model,
+        value_model,
+        model_based_env,
+        cfg,
+        value_loss: nn.Module = nn.MSELoss(),
+        gamma=0.99,
+        lmbda=0.95,
+    ):
+        super().__init__()
+        self.actor_model = actor_model
+        self.value_model = value_model
+        self.model_based_env = model_based_env
+        self.cfg = cfg
+        self.value_loss = value_loss
+        self.gamma = gamma
+        self.lmbda = lmbda
+
+    def forward(self, tensordict) -> torch.Tensor:
+        with torch.no_grad():
+            tensordict = tensordict.select("next_prior_state", "next_belief")
+
+            tensordict.batch_size = [
+                tensordict.shape[0],
+                tensordict.get("next_belief").shape[1],
+            ]
+            tensordict.rename_key("next_prior_state", "prior_state")
+            tensordict.rename_key("next_belief", "belief")
+            tensordict = tensordict.reshape(-1).detach()
+        with hold_out_net(self.model_based_env):
+            tensordict = self.model_based_env.rollout(
+                max_steps=self.cfg.imagination_horizon,
+                policy=self.actor_model,
+                auto_reset=False,
+                tensordict=tensordict,
+            )
+            with hold_out_net(self.value_model):
+                tensordict = self.value_model(tensordict)
+        lambda_target = self.lambda_target(
+            tensordict.get("reward"), tensordict.get("predicted_value")
+        )
+        actor_loss = -lambda_target.mean()
+        with torch.no_grad():
+            value_td = tensordict.clone().detach()
```

Comment:
Detach under no_grad?

Reply:
If not detach we have graph optimization overlap